File: apps/demo-be-rsax/tests/integration/steps/security_steps.rs
use cucumber::{given, then, when};

use crate::world::{json_req, AppWorld};

#[when(
    regex = r#"the client sends POST /api/v1/auth/register with body \{ "username": "alice", "email": "alice@example\.com", "password": "Short1!Ab" \}"#
)]
async fn register_short_password(world: &mut AppWorld) {
    let req = json_req(
        "POST",
        "/api/v1/auth/register",
        r#"{"username": "alice", "email": "alice@example.com", "password": "Short1!Ab"}"#,
        None,
    );
    world.send(req).await.unwrap();
}

#[when(
    regex = r#"the client sends POST /api/v1/auth/register with body \{ "username": "alice", "email": "alice@example\.com", "password": "AllUpperCase1234" \}"#
)]
async fn register_no_special_char(world: &mut AppWorld) {
    let req = json_req(
        "POST",
        "/api/v1/auth/register",
        r#"{"username": "alice", "email": "alice@example.com", "password": "AllUpperCase1234"}"#,
        None,
    );
    world.send(req).await.unwrap();
}

#[given(expr = "{string} has had the maximum number of failed login attempts")]
async fn max_failed_attempts(world: &mut AppWorld, username: String) {
    for _ in 0..5 {
        let body = format!(r#"{{"username": "{username}", "password": "WrongPass!123"}}"#);
        let req = json_req("POST", "/api/v1/auth/login", &body, None);
        world.send(req).await.unwrap();
    }
}

#[given(expr = "a user {string} is registered and locked after too many failed logins")]
async fn register_and_lock(world: &mut AppWorld, username: String) {
    let email = format!("{username}@example.com");
    let reg_body =
        format!(r#"{{"username": "{username}", "email": "{email}", "password": "Str0ng#Pass1"}}"#);
    let req = json_req("POST", "/api/v1/auth/register", &reg_body, None);
    world.send(req).await.unwrap();
    if world.last_status == 201 && username == "alice" {
        world.alice_id = world
            .last_body
            .get("id")
            .and_then(|v| v.as_str())
            .and_then(|s| uuid::Uuid::parse_str(s).ok());
    }

    for _ in 0..5 {
        let body = format!(r#"{{"username": "{username}", "password": "WrongPass!123"}}"#);
        let req = json_req("POST", "/api/v1/auth/login", &body, None);
        world.send(req).await.unwrap();
    }
}

#[given(expr = "an admin user {string} is registered and logged in")]
async fn register_admin_and_login(world: &mut AppWorld, username: String) {
    let email = format!("{username}@example.com");
    let reg_body =
        format!(r#"{{"username": "{username}", "email": "{email}", "password": "Str0ng#Pass1"}}"#);
    let req = json_req("POST", "/api/v1/auth/register", &reg_body, None);
    world.send(req).await.unwrap();

    if world.last_status == 201 {
        let admin_id = world
            .last_body
            .get("id")
            .and_then(|v| v.as_str())
            .and_then(|s| uuid::Uuid::parse_str(s).ok());

        if let Some(admin_uuid) = admin_id {
            world.promote_to_admin(admin_uuid).await.unwrap();
        }

        let login_body = format!(r#"{{"username": "{username}", "password": "Str0ng#Pass1"}}"#);
        let req = json_req("POST", "/api/v1/auth/login", &login_body, None);
        world.send(req).await.unwrap();
        if world.last_status == 200 {
            world.admin_token = world
                .last_body
                .get("access_token")
                .and_then(|v| v.as_str())
                .map(String::from);
        }
    }
}

#[given("an admin has unlocked alice's account")]
async fn admin_unlocked_alice(world: &mut AppWorld) {
    // Register an admin if we don't have one yet
    if world.admin_token.is_none() {
        let reg_body = r#"{"username": "sysadmin", "email": "sysadmin@example.com", "password": "Str0ng#Pass1"}"#;
        let req = crate::world::json_req("POST", "/api/v1/auth/register", reg_body, None);
        world.send(req).await.unwrap();
        if world.last_status == 201 {
            let admin_id = world
                .last_body
                .get("id")
                .and_then(|v| v.as_str())
                .and_then(|s| uuid::Uuid::parse_str(s).ok());
            if let Some(admin_uuid) = admin_id {
                world.promote_to_admin(admin_uuid).await.unwrap();
            }
            let login_body = r#"{"username": "sysadmin", "password": "Str0ng#Pass1"}"#;
            let req = crate::world::json_req("POST", "/api/v1/auth/login", login_body, None);
            world.send(req).await.unwrap();
            if world.last_status == 200 {
                world.admin_token = world
                    .last_body
                    .get("access_token")
                    .and_then(|v| v.as_str())
                    .map(String::from);
            }
        }
    }
    let admin_bearer = world.admin_bearer();
    let alice_id = world.alice_id.map(|id| id.to_string()).unwrap_or_default();
    if !alice_id.is_empty() && !admin_bearer.is_empty() {
        let req = crate::world::json_req(
            "POST",
            &format!("/api/v1/admin/users/{alice_id}/unlock"),
            "{}",
            Some(&admin_bearer),
        );
        world.send(req).await.unwrap();
    }
}

#[when("the admin sends POST /api/v1/admin/users/{alice_id}/unlock")]
async fn admin_unlock_alice(world: &mut AppWorld) {
    let admin_bearer = world.admin_bearer();
    let alice_id = world.alice_id.map(|id| id.to_string()).unwrap_or_default();
    let req = json_req(
        "POST",
        &format!("/api/v1/admin/users/{alice_id}/unlock"),
        "{}",
        Some(&admin_bearer),
    );
    world.send(req).await.unwrap();
}

File: apps/demo-be-java-vertx/src/test/java/com/organiclever/demojavx/integration/steps/SecuritySteps.java
package com.organiclever.demojavx.integration.steps;

import com.organiclever.demojavx.support.AppFactory;
import com.organiclever.demojavx.support.ServiceResponse;
import com.organiclever.demojavx.support.ScenarioState;
import io.cucumber.java.en.Given;
import io.cucumber.java.en.Then;
import io.cucumber.java.en.When;
import io.vertx.core.json.JsonArray;
import io.vertx.core.json.JsonObject;
import org.junit.jupiter.api.Assertions;

public class SecuritySteps {

    private static final int MAX_FAILED_ATTEMPTS = 5;
    private final ScenarioState state;

    public SecuritySteps(ScenarioState state) {
        this.state = state;
    }

    @Given("{string} has had the maximum number of failed login attempts")
    public void hasHadMaxFailedLoginAttempts(String username) throws Exception {
        for (int i = 0; i < MAX_FAILED_ATTEMPTS; i++) {
            AppFactory.getService().login(username, "WrongPassword!");
        }
    }

    @Then("alice's account status should be {string}")
    public void alicesAccountStatusShouldBe(String expectedStatus) throws Exception {
        String adminToken = state.getAdminAccessToken();
        String userId = state.getUserId();

        if (adminToken != null && userId != null) {
            ServiceResponse resp = AppFactory.getService()
                    .adminListUsers(adminToken, null, 1, 100);
            JsonObject body = resp.body();
            Assertions.assertNotNull(body);
            JsonArray data = body.getJsonArray("data");
            boolean found = false;
            for (int i = 0; i < data.size(); i++) {
                JsonObject user = data.getJsonObject(i);
                if (userId.equals(user.getString("id"))) {
                    Assertions.assertEquals(expectedStatus.toLowerCase(),
                            user.getString("status", "").toLowerCase());
                    found = true;
                    break;
                }
            }
            if (!found) {
                for (int i = 0; i < data.size(); i++) {
                    JsonObject user = data.getJsonObject(i);
                    if ("alice".equals(user.getString("username"))) {
                        Assertions.assertEquals(expectedStatus.toLowerCase(),
                                user.getString("status", "").toLowerCase());
                        return;
                    }
                }
            }
        } else {
            ServiceResponse lastResp = state.getLastResponse();
            Assertions.assertNotNull(lastResp);
            Assertions.assertEquals(401, lastResp.statusCode());
        }
    }

    @Given("a user {string} is registered and locked after too many failed logins")
    public void aUserIsRegisteredAndLocked(String username) throws Exception {
        String password = "Str0ng#Pass1";
        state.setPassword(password);
        AuthSteps authSteps = new AuthSteps(state);
        authSteps.registerUser(username, username + "@example.com", password);
        hasHadMaxFailedLoginAttempts(username);
    }

    @Given("an admin user {string} is registered and logged in")
    public void anAdminUserIsRegisteredAndLoggedIn(String username) throws Exception {
        AuthSteps authSteps = new AuthSteps(state);
        ServiceResponse regResp = authSteps.registerUser(username,
                username + "@example.com", "Admin#Pass1234");
        JsonObject regBody = regResp.body();
        Assertions.assertNotNull(regBody, "Registration response body must not be null");
        String adminId = regBody.getString("id");

        // Promote to admin directly (bypasses auth check)
        AppFactory.getService().promoteToAdmin(adminId);

        // Login as admin
        ServiceResponse loginResp = authSteps.login(username, "Admin#Pass1234");
        JsonObject loginBody = loginResp.body();
        Assertions.assertNotNull(loginBody);
        String adminToken = loginBody.getString("access_token");
        state.setAdminAccessToken(adminToken);
    }

    @Given("an admin has unlocked alice's account")
    public void anAdminHasUnlockedAlicesAccount() throws Exception {
        // Register and login a temporary admin
        AuthSteps authSteps = new AuthSteps(state);
        ServiceResponse regResp = authSteps.registerUser("tempAdmin",
                "tempAdmin@example.com", "Admin#Pass1234");
        JsonObject regBody = regResp.body();
        Assertions.assertNotNull(regBody);
        String adminId = regBody.getString("id");
        AppFactory.getService().promoteToAdmin(adminId);
        ServiceResponse loginResp = authSteps.login("tempAdmin", "Admin#Pass1234");
        JsonObject loginBody = loginResp.body();
        Assertions.assertNotNull(loginBody);
        String adminToken = loginBody.getString("access_token");

        // Find alice's user id
        ServiceResponse listResp = AppFactory.getService()
                .adminListUsers(adminToken, null, 1, 100);
        JsonObject listBody = listResp.body();
        Assertions.assertNotNull(listBody);
        String aliceId = findUserIdByUsername(listBody.getJsonArray("data"), "alice");

        // Unlock alice
        AppFactory.getService().adminUnlockUser(adminToken, aliceId);
    }

    @When("^the admin sends POST /api/v1/admin/users/\\{alice_id\\}/unlock$")
    public void adminSendsUnlock() throws Exception {
        String adminToken = state.getAdminAccessToken();
        Assertions.assertNotNull(adminToken);

        ServiceResponse listResp = AppFactory.getService()
                .adminListUsers(adminToken, null, 1, 100);
        JsonObject listBody = listResp.body();
        Assertions.assertNotNull(listBody);
        String aliceId = findUserIdByUsername(listBody.getJsonArray("data"), "alice");

        ServiceResponse response = AppFactory.getService().adminUnlockUser(adminToken, aliceId);
        state.setLastResponse(response);
    }

    public static String findUserIdByUsername(JsonArray data, String username) {
        for (int i = 0; i < data.size(); i++) {
            JsonObject user = data.getJsonObject(i);
            if (username.equals(user.getString("username"))) {
                return user.getString("id");
            }
        }
        return "";
    }
}
